docs/src/index.md

```diff
@@ -6,4 +6,369 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+    And there exact functioning is kind of ChainRule specific,
+    though other tools may do similar.
+    The core notion is sometimes called
+    _Custom AD primitives_, _custom adjoints_, _custom sensitivities_.
+
+The rules are encoded as `rrules` and `frules`,
+for use in reverse-mode and forward-mode differentiation respectively.
+
+The `rrule` for some function `foo`, takes the positional argument `args` and keyword argument `kwargs` is written:
+```julia
+function rrule(::typeof(foo), args; kwargs...)
+    ...
+    return y, pullback
+end
+```
+where `y` must be equal to `foo(args; kwargs...)`,
+and _pullback_ is a function to propagate the derivative information backwards at that point (more later).
+Often but not always it is calculated directly.
+The exception is that we can calculate it indirectly to make
+the `pullback` faster. (more on _pullback_ later)
+
+Similarly, the `frule` is written:
+```julia
+function frule(::typeof(foo), args; kwargs...)
+    ...
+    return y, pushforward
+end
+```
+again `y=foo(args, kwargs...)`,
+and _pushforward_ is a function to propagate the derivative information forwards at that point (more later).
+
+Almost always the _pushforward_/_pullback_ will be declared locally with-in the `ffrule`/`rrule`, and will be a _closure_ over some of the other arguments.
+
+### The propagators: pushforward and pullback
+
+!!! Terminology "Pushforward and Pullback"
+
+    _Pushforward_ and _Pullback_ are fancy words that the autodiff community recently stole from Differential Geometry.
+    The are broadly in agreement with the use of these terms in differential geometry. But any geometer will tell you these are the super-boring flat cases. Some will also frown at you.
+    Other terms that may be used include for _pullback_ the **backpropagator**, and by analogy for _pushforward_ the **forwardpropagator**, thus these are the _propagators_.
+    These are also good names because effectively they propagate wibbles and wobbles through them, via the chainrule.
+    (the term **backpropagator** may originate with ["Lambda The Ultimate Backpropagator"](http://www-bcl.cs.may.ie/~barak/papers/toplas-reverse.pdf) by Bearlmutter and Siskind, 2008)
+
+
+#### Core Important Idea:
+ - The **Pushforward** takes a wiggle in the _input space_, and tells what wobble you would create in the output space, by passing it through the function.
+ - The **Pullback** takes a wobble in the _output space_, and tells you what wiggle you would need to make in the _input_ space to achieve it.
+
+#### The anatomy of pushforward and pullback
+
+For our function `foo(args...; kwargs) = Y`:
+
+The pushforward is a function:
+```julia
+function pushforward(Δself, Δargs...)
+    ...
+    return ∂Y
+end
+```
+**Note:** that there is one `Δargs...` per `arg` to the original function, and they are similar in type/structure to the corresponding inputs.
+Plus the `Δself` (don't worry we will be back to explain this soon).
```

Comment:
>Plus the `Δself` (don't worry we will be back to explain this soon).

Don't say 'plus' when you're talking about the inclusion of things that could reasonably be added together.

I would just say, "the presence of self will be explained soon".

Reply:
Good point

Reply:
```suggestion
as well as the `Δself` (the presence of `Δself` will be explained soon).
```